FUSE: increase code coverage

StefanoDuo opened the pull request:
This patch set increases the total FUSE code coverage by adding a pseudo-syscalls (and some other stuff) which helps syzkaller mimics a FUSE daemon behavior.

Note: `make presubmit` fails at the `make lint` with the following error message `pkg/host/syscalls_linux.go:173:1: cyclomatic complexity 27 of func "isSupportedSyzkall" is high (> 24)`. This seems to be due to the additional branches in the "isSupportedSyzkall", does anyone have any solution for that?
I have run `make test`, `make presubmit_arch` and `make presubmit_race` manually and they pass.

---
CC: @balsini

xairy commented on executor/common_linux.h:
```diff
@@ -4103,3 +4103,281 @@ static void setup_usb()
+ * 2) Mount FUSE fs on top of target directory
+ * 3) Return fd referring to the target directory
+ */
+static volatile long syz_mount_fuse(volatile long a0, /* /dev/fuse fd. */
```
Please prefix all FUSE pseudo-syscalls with `syz_fuse_`. E.g. `syz_fuse_mount` instead of `syz_mount_fuse`, etc.

StefanoDuo replied:
Done.

dvyukov commented:
> Note: `make presubmit` fails at the `make lint` with the following error message `pkg/host/syscalls_linux.go:173:1: cyclomatic complexity 27 of func "isSupportedSyzkall" is high (> 24)`. This seems to be due to the additional branches in the "isSupportedSyzkall", does anyone have any solution for that?

Hi Stefano,

It needs to be refactored somehow to reduce cyclomatic complexity.
Perhaps a map from syscall name to checking function will help to reduce it once and for all.
An intermediate solution may be to combine some functions into a single case, but I am not sure how exactly the checker counts, maybe it won't help.

dvyukov commented on executor/common_linux.h:
```diff
@@ -4103,3 +4103,281 @@ static void setup_usb()
+	dir_fd = (int)a0;
+	filename = (char*)a1;
+	open_flags = ((int)a2) | O_CREAT;
+	fd = openat(dir_fd, filename, open_flags);
```
Pseudo-syscalls is a last resort cases that we can't handle in descriptions and even can't handle any approximation of what we need.
This pseudo-syscall does not seem to do anything that can't be done in descriptions, use openat syscall specialization in descriptions instead.

StefanoDuo replied:
Is it possible to force the O_CREAT flag while fuzzing the remaining flags?

I had initially tested using a specialized syscall and this didn't seem to be the case (and fuzzing the open flag without forcing O_CREAT lead to a worse code coverage).

dvyukov replied:
See `syzkaller$ grep "define .*|" sys/linux/*.txt`
But why do we want O_CREAT always? I think we want to open existing files equally, no?

> and fuzzing the open flag without forcing O_CREAT lead to a worse code coverage

Testing fuzzer changes is tricky. What was the methodology for testing?

StefanoDuo replied:
> But why do we want O_CREAT always? I think we want to open existing files equally, no?

O_CREAT just says create if missing, it still allows opening existing files. To test FUSE, we are mounting a brand new filesystem every time and therefore almost always openat() is targeting nonexistent files (this is why I would like to always set O_CREAT for this specific use case).

> Testing fuzzer changes is tricky. What was the methodology for testing?

I have come to realize that, haha. The best I have come up with is letting it run for 1-3 days on 12 VMs with 2 CPUs each (~ 300-900 hours of fuzzing). Is there a better way?

dvyukov replied:
Right, I confused it with O_EXCL.
Every test starts with an empty dir every time, so this looks like non-fuse specific and affects all open calls.

> I have come to realize that, haha. The best I have come up with is letting it run for 1-3 days on 12 VMs with 2 CPUs each (~ 300-900 hours of fuzzing). Is there a better way?

No :) But at least it's not a 1-hour run.

StefanoDuo replied:
> I had initially tested using a specialized syscall and this didn't seem to be the case (and fuzzing the open flag without forcing O_CREAT lead to a worse code coverage).

After testing again, files are being opened correctly using the normal openat() syscall (as one would expect). I think I was defining the openat$fuse_file(...) variant with 3 parameters (no mode) and that was somehow messing things up.

dvyukov commented on executor/common_linux.h:
```diff
@@ -4103,3 +4103,281 @@ static void setup_usb()
+	mount_target = (char*)a1;
+	mount_flags = (unsigned long)a2;
+
+	ret = mkdir(mount_target, S_IRWXU);
```
I don't see what's here cannot be done in descriptions. Am I missing something? Have you tried to express all of this in descriptions? 
See sys/linux/filesystem.txt for now to form mount options.

StefanoDuo replied:
The important bit about this syscalls is the returned file descriptor. It refers to the directory where the FUSE filesystem has been mounted, and allows the syscalls that follow to target files stored on a FUSE filesystem without hardcoding paths (using `*at()` syscalls and passing this fd). From what I have seen, there isn't a real syscall that supports this behavior.

Regarding fuzzing the mount options, a discriminated syscall for FUSE is already present inside [sys/linux/fs_fuse.txt](https://github.com/google/syzkaller/blob/master/sys/linux/fs_fuse.txt#L16).

dvyukov replied:
We don't need a pseudo-syscall for this. An openat specialization can accept and return a more specialized fd type.

dvyukov replied:
Oh, wait, I was thinking about something else (the other pseudo-syscall).

dvyukov replied:
I still don't see what we gain much here. All scenarios that are possible with these pseudo-syscalls seem to be easily possible with mkdir/openat as well. It maybe will shift the focus just a bit, but it does not seem to justify more custom C code, more maintainable, larger reproducers. "Slightly shifting focus" may be useful for all of 10000 kernel subsystems, but we can't add 10000 pseudo-syscalls just for that.
If you really want to improve this, we need to think how syzlang needs to be changed to support such richer resource annotations.

StefanoDuo replied:
I have tried and wasn't able to reproduce FUSE file interaction without this and syz_handle_fuse_req() (with this being strictly needed for interacting with directories).

My guess it that without returning the fd from the mount() syscall, syzkaller is free to execute the mount() and mkdir()/openat() on 2 different threads since there isn't any dependency between the two syscalls. This leads to a data race, and can end up with mkdir()/openat() executing before mount() and targeting the previously mounted filesystem (e.g. ext4). Please, correct me if what I'm saying is wrong.

StefanoDuo replied:
Coming back to this after a few days of fuzzing. Using only mkdir/openat$dir/mount$fuse, the fuzzer hasn't been able to cover all the code it was covering with the pseudo-syscall syz_fuse_mount(). The main problem seems to be opening "FUSE directories" because `fs/fuse/readdir.c` hasn't been touched at all (previously ~72% coverage).

If you are still against this pseudo-syscall, I can generalize it to something like syz_mount(), accepting the same parameters as a real mount(), but returning the mounted directory fd. This way, all filesystems can use it, and for FUSE I can hardcode the parameters needed in a syscall specialization. Would that be acceptable?

dvyukov replied:
Interesting. I wonder what's the reason for no coverage in fs/fuse/readdir.c.
I am now trying to re-create the experiment.

dvyukov replied:
I also need to enable openat$fuse

dvyukov replied:
also these:
```
		"read$FUSE",
		"write$FUSE*",
		"syz_fuse_handle_req",
		"getuid",
```

dvyukov replied:
Did you enable getdents? That's what I think will give coverage in fs/fuse/readdir.c.

StefanoDuo replied:
> Coming back to this after a few days of fuzzing. Using only mkdir/openat$dir/mount$fuse, the fuzzer hasn't been able to cover all the code it was covering with the pseudo-syscall syz_fuse_mount().

Sorry, I didn't express myself too well. With "Using only mkdir/openat$dir/mount$fuse" I meant these instead of syz_fuse_mount (combined by the fuzzer they should substitute syz_fuse_mount). You still need to enable VFS/FUSE related syscalls to reproduce some interaction.

StefanoDuo replied:
> Did you enable getdents? That's what I think will give coverage in fs/fuse/readdir.c.

Yes. Just to be sure, I am rerunning syzkaller with the following config:
```
"disable_syscalls": [
                "fsetxattr$security_capability",
                "fsetxattr$security_evm",
                "fsetxattr$security_ima",
                "fsetxattr$security_selinux",
                "fsetxattr$system_posix_acl",
                "fsetxattr$trusted_overlay_nlink",
                "fsetxattr$trusted_overlay_opaque",
                "fsetxattr$trusted_overlay_origin",
                "fsetxattr$trusted_overlay_redirect",
                "fsetxattr$trusted_overlay_upper",
                "mkdirat$cgroup",
                "mkdirat$cgroup_root",
                "openat$autofs",
                "openat$bsg",
                "openat$cgroup",
                "openat$cgroup_devices",
                "openat$cgroup_freezer_state",
                "openat$cgroup_int",
                "openat$cgroup_netprio_ifpriomap",
                "openat$cgroup_procs",
                "openat$cgroup_ro",
                "openat$cgroup_root",
                "openat$cgroup_subtree",
                "openat$cgroup_type",
                "openat$full",
                "openat$hpet",
                "openat$hwrng",
                "openat$loop_ctrl",
                "openat$md",
                "openat$null",
                "openat$nvram",
                "openat$pidfd",
                "openat$procfs",
                "openat$ptmx",
                "openat$ptp0",
                "openat$random",
                "openat$rfkill",
                "openat$rtc",
                "openat$selinux_attr",
                "openat$snapshot",
                "openat$sndseq",
                "openat$sndtimer",
                "openat$thread_pidfd",
                "openat$ttyS3",
                "openat$urandom",
                "openat$vcs",
                "openat$vcsa",
                "openat$vcsu",
                "openat$vga_arbiter",
                "openat$zero",
                "read$ptp",
                "read$rfkill",
                "read$snapshot",
                "write$binfmt_aout",
                "write$binfmt_elf32",
                "write$binfmt_elf64",
                "write$binfmt_misc",
                "write$binfmt_script",
                "write$cgroup_devices",
                "write$cgroup_freezer_state",
                "write$cgroup_int",
                "write$cgroup_netprio_ifpriomap",
                "write$cgroup_pid",
                "write$cgroup_subtree",
                "write$cgroup_type",
                "write$rfkill",
                "write$selinux_attr",
                "write$snapshot",
                "write$sndseq"
        ],
        "enable_syscalls": [
                "openat$fuse",
                "openat$cuse",
                "openat$dir",
                "openat",
                "close",
                "mount$fuse",
                "mount$fuseblk",
                "umount2",
                "ioctl$FUSE_DEV_IOC_CLONE",
                "ioctl$FIBMAP",
                "write",
                "read",
                "syz_fuse_handle_req",
                "mkdirat",
                "mkdir",
                "renameat",
                "ftruncate",
                "fallocate",
                "syncfs",
                "fdatasync",
                "getdents",
                "getdents64",
                "readlinkat",
                "unlinkat",
                "symlinkat",
                "linkat",
                "sync_file_range",
                "splice",
                "poll",
                "fsync",
                "fchmod",
                "fstatfs",
                "fstat",
                "fgetxattr",
                "fsetxattr",
                "flistxattr",
                "fremovexattr",
                "mmap",
                "lseek",
                "flock",
                "fcntl$lock",
                "getgid",
                "getuid",
                "getpid"
        ]
```

StefanoDuo replied:
> Just to be sure, I am rerunning syzkaller with the following config: [...]

Reconfirming that after ~260 hours of fuzzing `fs/fuse/readdir.c` has no coverage.

dvyukov replied:
Weird. I hate writing this off to "fuzzer not being able to open the dir", because it's really just 1 additional openat call. There may be some more fundamental issue here. But I got distracted from chasing it further yesterday...

StefanoDuo replied:
> Weird. I hate writing this off to "fuzzer not being able to open the dir", because it's really just 1 additional openat call.

I agree with that, but currently I don't know why it is not happening. To make sure getdents() can be called, I have written this syzkaller program (which does not use syz_fuse_mount()), and it works:
```
mkdir(&(0x7f0000000100)='./file0\x00', 0x0)
r0 = openat$fuse(0xffffffffffffff9c, &(0x7f0000000000)='/dev/fuse\x00', 0x2, 0x0)
mount$fuse(0x0, &(0x7f0000000040)='./file0\x00', &(0x7f0000000080)='fuse\x00', 0x0, &(0x7f0000000bc0)={{'fd', 0x3d, r0}, 0x2c, {'rootmode', 0x3d, 0x4000}, 0x2c, {'user_id'}, 0x2c, {'group_id'}})
read$FUSE(r0, &(0x7f0000000c80)={0x2020, 0x0, <r1=>0x0}, 0x2020)
write$FUSE_INIT(r0, &(0x7f0000000140)={0x50, 0x0, r1}, 0x50)
r3 = openat$dir(0xffffffffffffff9c, &AUTO='./file0\x00', 0x0, 0x0)
syz_fuse_handle_req(r0, &(0x7f00000061c0)=""/8192, 0x2000, &(0x7f0000000c00)={0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, &(0x7f0000000100)={0x20, 0x0, 0x0, {0x0, 0x8}}, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0})
getdents(r3, &AUTO=""/1024, AUTO)
syz_fuse_handle_req(r0, &(0x7f0000006180)=""/8192, 0x2000, &(0x7f0000000880)={0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, &(0x7f00000005c0)={0x30, 0x0, 0x0, [{0x0, 0x0, 0x4, 0x0, '#,,-'}]}, 0x0, 0x0, 0x0})
```

The syscalls I use are all enabled in the above configs, so the fuzzer should be able to come up with this.

Taking a quick glance at the programs generated by the fuzzer, I have noticed that when it calls opeat$dir() it only tries to open nested directories and not the FUSE root dir (and this requires a lot more exchanges between FUSE kernel and FUSE daemon). Maybe that is related to the issue.

dvyukov replied:
Running with these enabled syscalls:
```
	"enable_syscalls": [
		"mount$fuse",
		"mkdirat",
		"openat",
		"write",
		"read",
		"getdents",
		"openat$fuse",
		"openat$dir",
		"read$FUSE",
		"write$FUSE*",
		"syz_fuse_handle_req",
		"getuid"
	],
```
for:
```
uptime | 20h2m10s
fuzzing | 588h30m0s
```
I've got 41% coverage in fs/fuse/readdir.c

dvyukov replied:
I've tried to enforce O_CREAT for all openat's with:
```
--- a/sys/linux/init.go
+++ b/sys/linux/init.go
@@ -14,6 +14,7 @@ func InitTarget(target *prog.Target) {
        arch := &arch{
                unix:                        targets.MakeUnixNeutralizer(target),
                clockGettimeSyscall:         target.SyscallMap["clock_gettime"],
+               O_CREAT:                     target.GetConst("O_CREAT"),
                MREMAP_MAYMOVE:              target.GetConst("MREMAP_MAYMOVE"),
                MREMAP_FIXED:                target.GetConst("MREMAP_FIXED"),
                SYSLOG_ACTION_CONSOLE_OFF:   target.GetConst("SYSLOG_ACTION_CONSOLE_OFF"),
@@ -113,6 +114,7 @@ type arch struct {
 
        clockGettimeSyscall *prog.Syscall
 
+       O_CREAT                     uint64
        MREMAP_MAYMOVE              uint64
        MREMAP_FIXED                uint64
        SYSLOG_ACTION_CONSOLE_OFF   uint64
@@ -149,6 +151,8 @@ type arch struct {
 func (arch *arch) neutralize(c *prog.Call) {
        arch.unix.Neutralize(c)
        switch c.Meta.CallName {
+       case "openat":
+               c.Args[2].(*prog.ConstArg).Val |= arch.O_CREAT
        case "mremap":
                // Add MREMAP_FIXED flag, otherwise it produces non-deterministic results.
                flags := c.Args[3].(*prog.ConstArg)
```

but after I restarted the same instances, all coverage in readdir.c disappeared (though, majority of coverage in fuse stayed). Not sure if it's something fundamental, or it just needs to create new reproducers that will work with O_CREAT.

dvyukov replied:
After reverting the O_CREAT patch, I get 46% readdir.c coverage within 1.5h.

StefanoDuo replied:
> Running with these enabled syscalls:
> 
> ```
> 	"enable_syscalls": [
> 		"mount$fuse",
> 		"mkdirat",
> 		"openat",
> 		"write",
> 		"read",
> 		"getdents",
> 		"openat$fuse",
> 		"openat$dir",
> 		"read$FUSE",
> 		"write$FUSE*",
> 		"syz_fuse_handle_req",
> 		"getuid"
> 	],
> ```
> 
> for:
> 
> ```
> uptime | 20h2m10s
> fuzzing | 588h30m0s
> ```
> 
> I've got 41% coverage in fs/fuse/readdir.c

I have checked my instance over the weekend as well, and confirm that I had ~43% coverage in `fs/fuse/readdir.c` as well after ~600 hours of fuzzing.

> After reverting the O_CREAT patch, I get 46% readdir.c coverage within 1.5h.

Is the fuzzer reusing the previously dropped programs (due to the O_CREAT patch introduction)?

dvyukov replied:
> > After reverting the O_CREAT patch, I get 46% readdir.c coverage within 1.5h.
> 
> Is the fuzzer reusing the previously dropped programs (due to the O_CREAT patch introduction)?

Hard to say.
After each restart it triages all programs in the corpus and throws away the ones that don't give new coverage. So it could throw away lots of them. However, it could keep them if they give new coverage in other parts of the kernel. So maybe it preserved just enough to regain majority of coverage in readdir.

dvyukov commented on sys/linux/test/fuse_deadlock:
```diff
@@ -4,8 +4,8 @@
 r0 = openat$fuse(0xffffffffffffff9c, &AUTO='/dev/fuse\x00', 0x2, 0x0)
 mount$fuse(0x0, &AUTO='./file0\x00', &AUTO='fuse\x00', 0x0, &AUTO={{'fd', 0x3d, r0}, 0x2c, {'rootmode', 0x3d, 0x4000}, 0x2c, {'user_id', 0x3d, 0x0}, 0x2c, {'group_id', 0x3d, 0x0}, 0x2c, {[], [], 0x0}})
-read$FUSE(r0, &AUTO=""/4096, AUTO)
+read$FUSE(r0, &AUTO={AUTO, 0x0, <r1=>0x0, 0x0, 0x0, 0x0, 0x0, ""/8192}, AUTO)
```
I wonder how it was working before with 4096... do you have any hypothesis?

StefanoDuo replied:
I haven't found any documentation on that, but looking at the [source code](https://github.com/torvalds/linux/blob/master/fs/fuse/dev.c#L1199) it seems that if the total read buffer length is less than FUSE_MIN_READ_BUFFER, a read on /dev/fuse returns EINVAL.

Was it working without returning any error (I haven't tested this specific script, only a C program interacting with /dev/fuse directly)?

dvyukov commented on sys/linux/fs_fuse.txt:
```diff
@@ -50,13 +59,31 @@ type fuse_notify[MSG, PAYLOAD] {
+type fuse_out[PAYLOAD] _fuse_out[fuse_unique, PAYLOAD]
+# This response header is used by syz_fuse_handle_req(). It defines the FUSE unique
+# identifier as int64 because syz_fuse_handle_req() retrieves it internally
+# (defining it as a resource would create a dependency with read$fuse() which is
```
s/read$fuse/read$FUSE/

StefanoDuo replied:
Done.

dvyukov commented on sys/linux/fs_fuse.txt:
```diff
@@ -142,8 +169,12 @@ fuse_write_out {
+
 fuse_open_out {
-	fh		const[0, int64]
+	fh		int64
```
Why?

StefanoDuo replied:
From my understanding, fh will be used by the kernel to refer to the resource being opened. It is supplied by the FUSE daemon and there are no restriction on its specific value. Therefore, in my opinion, it makes more sees not to hardcode it to 0. What do you think?

dvyukov replied:
Does kernel use it in any way? From briefly skimming through the kenrel code, it looks like an opaque handle for userspace.

StefanoDuo replied:
I haven't checked, but it probably doesn't. Still, this might help catch bugs that can be introduced in the future.

dvyukov replied:
This is exactly the opposite argument you make with the pseudo-syscalls ;)

dvyukov replied:
Also semantics of this field can't change in future, it's part of stable kernel API with lots of existing programs passing arbitrary data in this field already.

StefanoDuo replied:
I see your point, I have no problems dropping the change. Thanks for pointing that out!

StefanoDuo commented on pkg/host/syscalls_linux.go:
```diff
@@ -238,6 +238,12 @@ func isSupportedSyzkall(c *prog.Syscall, target *prog.Target, sandbox string) (b
 			return false, fmt.Sprintf("sys_%v is not present in the target", ioUringSyscallName)
 		}
 		return isSupportedSyscall(ioUringSyscall, target)
+	case "syz_mount_fuse":
```
Done.

StefanoDuo commented on pkg/host/syscalls_linux.go:
```diff
@@ -238,6 +238,12 @@ func isSupportedSyzkall(c *prog.Syscall, target *prog.Target, sandbox string) (b
 		}
 		return isSupportedSyscall(ioUringSyscall, target)
+	case "syz_mount_fuse":
+		return true, ""
```
I tested and it fails under setuid sandbox, probably due to permission on /dev/fuse (I think they are r/w only for root by default). I updated the checks accordingly.

StefanoDuo commented on sys/linux/fs_fuse.txt:
```diff
@@ -50,13 +59,31 @@ type fuse_notify[MSG, PAYLOAD] {
 	payload	PAYLOAD
 } [packed]
 
+type _fuse_out[UNIQUE, PAYLOAD] {
```
Done.

dvyukov commented:
My local instance running without syz_fuse_mount for 20h more now got to 61% of readdir.c. So it seems that the fuzzer is capable of getting all that coverage, maybe just somewhat slower. Slightly bumping priority of using the directory with a new pseudo-syscall does not look like a good RoI. And this helps only 1 filesystem out of 92, while applicable to all of them. I am thinking about some alternative approaches.
1. It seem that we could slightly modify existing syz_mount_image to get the same effect: pass empty image for fuse, and then if image is empty, skip the memfd/loop part and just do mkdir+mount and add return of the dir fd. This looks equivalent and will benefit all filesystems with few additional lines of C code.
2. Preseeding corpus  with more complex scenarios seems to be useful. Please add the getdents program to sys/linux/test as well and we will figure out how to inject them into fuzzer corpus. If we start injecting them, then more non-trivial programs there will be useful. I hate working around fuzzer deficiencies with brute-force preseeding, but, well, all is fair in love, war and fuzzing.
3. Perhaps we can do some generic improvement for O_CREAT that will benefit all operations on all filesystems. But I am not sure what exactly it should be... any ideas?
4. Perhaps we can do some generic improvement in how the fuzzer works with files/directories. For example, how it chooses existing/new filename:
https://github.com/google/syzkaller/blob/70301872e129e968c3027d181efb25fce3de8707/prog/rand.go#L278-L302
or how fd/fd_dir are organized. fd_dir is a bit of hack now. It does not really anyhow biased towards opening directories, and e.g. we may already have an fd that refers to a dir, but then if we need an fd_dir that existing fd won't be used. As far as I remember I introduced fd_dir just to mark some arguments as "in this position we want a directory". Maybe a better split would be fd and fd_file (refers to some file/dir rather than pipe/dev/memfd/epoll/etc). Or maybe we need to change filenameImpl to use fileN and dirN elements so that it knows later what is what. But I did not think much about this, I am not sure if it will work out or not.

StefanoDuo commented:
> My local instance running without syz_fuse_mount for 20h more now got to 61% of readdir.c. So it seems that the fuzzer is capable of getting all that coverage, maybe just somewhat slower. Slightly bumping priority of using the directory with a new pseudo-syscall does not look like a good RoI. And this helps only 1 filesystem out of 92, while applicable to all of them. I am thinking about some alternative approaches.
> 
> 1. It seem that we could slightly modify existing syz_mount_image to get the same effect: pass empty image for fuse, and then if image is empty, skip the memfd/loop part and just do mkdir+mount and add return of the dir fd. This looks equivalent and will benefit all filesystems with few additional lines of C code.

I was thinking that a lot more was going on inside syz_mount_image, but I guess that is not the case. I have refactored it and added what I was previously doing inside syz_fuse_mount.

> 2. Preseeding corpus  with more complex scenarios seems to be useful. Please add the getdents program to sys/linux/test as well and we will figure out how to inject them into fuzzer corpus. If we start injecting them, then more non-trivial programs there will be useful. I hate working around fuzzer deficiencies with brute-force preseeding, but, well, all is fair in love, war and fuzzing.

Agreed.

dvyukov commented on executor/common_linux.h:
```diff
@@ -2509,50 +2544,42 @@ static long syz_read_part_table(volatile unsigned long size, volatile unsigned l
 // }
 static long syz_mount_image(volatile long fsarg, volatile long dir, volatile unsigned long size, volatile unsigned long nsegs, volatile long segments, volatile long flags, volatile long optsarg)
```
Update return type in the comment above.

dvyukov commented on sys/linux/fs_fuse.txt:
```diff
@@ -39,9 +41,11 @@ write$FUSE_NOTIFY_STORE(fd fd_fuse, arg ptr[in, fuse_notify[FUSE_NOTIFY_STORE, f
 write$FUSE_NOTIFY_RETRIEVE(fd fd_fuse, arg ptr[in, fuse_notify[FUSE_NOTIFY_RETRIEVE, fuse_notify_retrieve_out]], len bytesize[arg])
 write$FUSE_NOTIFY_DELETE(fd fd_fuse, arg ptr[in, fuse_notify[FUSE_NOTIFY_DELETE, fuse_notify_delete_out]], len bytesize[arg])
 
+syz_mount_image$fuse(fs ptr[in, string["fuse"]], dir ptr[in, filename], size const[0], nsegs const[0], segments const[0], flags flags[mount_flags], opts ptr[in, fuse_options]) fd_dir
```
Also update all other specializations of syz_mount_image to return fd_dir.

StefanoDuo replied:
Done.

dvyukov commented on executor/common_linux.h:
```diff
@@ -2393,14 +2393,13 @@ struct fs_image_segment {
 	// But some filesystems have large number of segments (2000+),
 	// we can't allocate that much on stack and allocating elsewhere is problematic,
 	// so we just use the memory allocated by fuzzer.
-	struct fs_image_segment* segs = (struct fs_image_segment*)segments;
```
While we are here: please also remove the comment above.
It's unrelated to your change, but we don't use NONFAILING in pseudo-syscalls at all, so it's not relevant anymore.

dvyukov commented on executor/common_linux.h:
```diff
@@ -2417,17 +2416,24 @@ static unsigned long fs_image_segment_check(unsigned long size, unsigned long ns
+	}
+
+	size = fs_image_segment_check(size, nsegs, segs);
+	memfd = syscall(sys_memfd_create, memfd_name, 0);
```
I don't think this memfd_name is important to preserve, it's not used for anything meaningful. It just used to be different because we had code duplicated. But now it does not worth the additional clutter in all reproducers (most of which will probably use only one of the syscalls anyway). So just use something like "syzkaller" or something.

dvyukov commented on executor/common_linux.h:
```diff
@@ -2417,17 +2416,24 @@ static unsigned long fs_image_segment_check(unsigned long size, unsigned long ns
+// creating and initializing the underlying file backing the loop device and
+// returns the fds to the file and device.
+// Returns 0 on success, -1 otherwise.
+static int setup_loop_device(long unsigned size, long unsigned nsegs, struct fs_image_segment* segs, const char* memfd_name, const char* loopname, int* _memfd, int* _loopfd)
```
We don't have any other precedents of identifiers starting with underscores. Please keep the code consistent. Start it with p or end with _p or something similar.
One other option is to return loop fd as the function result (it's called setup_loop_device in the end).

StefanoDuo replied:
I think the 2 pointers are more intuitive since we must return both memfd and loop file descriptors. I'll update their names to maintain consistency with the rest of code.

dvyukov commented on executor/common_linux.h:
```diff
@@ -2417,17 +2416,24 @@ static unsigned long fs_image_segment_check(unsigned long size, unsigned long ns
+	int loopfd;
+	int memfd;
+
+	if (!loopname || !memfd_name || !_memfd || !_loopfd) {
```
We don't do this level of defensive programming.
We have tests and this clutters C reproducers.
memfd_name you pass as non-NULL just a few lines below. The same for loopname. Checking every variable for NULL every other line is just not worth it.
And if you are thinking about fuzzer-provided pointers, you can't defend against them. It won't give you NULL, it will give you 0x555555555555555555 ;)

dvyukov commented on executor/common_linux.h:
```diff
@@ -2509,50 +2544,42 @@ static long syz_read_part_table(volatile unsigned long size, volatile unsigned l
-		if (pwrite(memfd, segs[i].data, segs[i].size, segs[i].offset) < 0) {
-			debug("syz_mount_image: pwrite[%u] failed: %d\n", (int)i, errno);
+	char loopname[64], opts[256];
+	char* source = NULL;
```
You need to include stddef.h to use NULL.

dvyukov commented on executor/common_linux.h:
```diff
@@ -2509,50 +2544,42 @@ static long syz_read_part_table(volatile unsigned long size, volatile unsigned l
+		// filesystem image.
+		memset(loopname, 0, sizeof(loopname));
+		snprintf(loopname, sizeof(loopname), "/dev/loop%llu", procid);
+		if (setup_loop_device(size, nsegs, segs, "syz_mount_image", loopname, &memfd, &loopfd) == -1) {
```
Drop {} around 1-statement block.

dvyukov commented on executor/common_linux.h:
```diff
@@ -2509,50 +2544,42 @@ static long syz_read_part_table(volatile unsigned long size, volatile unsigned l
-	strncpy(fs, (char*)fsarg, sizeof(fs) - 1);
+	mkdir(target, 0777);
 	memset(opts, 0, sizeof(opts));
-	// Leave some space for the additional options we append below.
```
Restore the comment.

StefanoDuo replied:
The comment is still there, I have just moved it up a little where we check if are truncating the received mount options. I don't think this will ever happen but having a debug in case that happens sounds ok to me, what do you think?

dvyukov commented on executor/common_linux.h:
```diff
@@ -4285,3 +4318,194 @@ static void setup_usb()
+	req_out = (struct syz_fuse_req_out*)a3;
+
+	if (!req_out) {
+		debug("syz_fuse_handle_req(): Received a NULL syz_fuse_req_out\n");
```
Please keep message capitalization consistent. Few lines above you use non-capital after `:`:
```
debug("fuse_send_response(): received a NULL out_hdr\n");
```
We generally don't use capital letter in beginning on log message.
Also we don't write `()` after function names.

StefanoDuo replied:
Done.

dvyukov commented on executor/common_linux.h:
```diff
@@ -4285,3 +4318,194 @@ static void setup_usb()
+		debug("syz_fuse_handle_req > read() failed: %d\n", errno);
+		return -1;
+	}
+	// Safe to do beacuse ret > 0 (!= -1) and < FUSE_MIN_READ_BUFFER
```
We don't  use 80-colon typewriters to work with this code ;)
Don't leaving just few characters on the next line will make it tidier:
```
// Safe to do beacuse ret > 0 (!= -1) and < FUSE_MIN_READ_BUFFER (= 8192).
```
(few lines below you violate 80 chars anyway)

StefanoDuo replied:
I think the comment was initially longer, and then I removed something. Anyway I have updated it.

StefanoDuo commented on pkg/host/syscalls_linux.go:
```diff
@@ -250,6 +250,24 @@ func isBtfVmlinuxSupported(c *prog.Syscall, target *prog.Target, sandbox string)
+	if err := osutil.IsAccessible("/dev/fuse"); err != nil {
+		return false, err.Error()
+	}
+	fuseSyscallName := "openat$fuse"
```
Done.

dvyukov commented:
The change looks good to me.
Please rebase to resolve conflicts. We stepped onto each other, sorry. `go test ./executor` will be a fast way to test for new warnings.

StefanoDuo commented:
> Please rebase to resolve conflicts. We stepped onto each other, sorry. `go test ./executor` will be a fast way to test for new warnings.

All conflicts should be resolved now. Thanks for suggesting `go test ./executor`, it definitely helped to speed up the process.
